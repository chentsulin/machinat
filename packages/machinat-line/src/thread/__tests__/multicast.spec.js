import Machinat from 'machinat';
import MulticastThread from '../multicast';

it('implements MachinatThread', () => {
  const thread = new MulticastThread(['foo', 'bar', 'baz']);

  expect(thread.platform).toBe('line');
  expect(thread.type).toBe('multicast');
  expect(thread.allowPause).toBe(false);
  expect(thread.to).toEqual(['foo', 'bar', 'baz']);
});

describe('#createJobs(actions)', () => {
  const Foo = () => {};

  const Bar = () => {};
  Bar.$$entry = () => 'bar';
  Bar.$$hasBody = false;

  const actions = [
    { element: <Foo />, value: { id: 0 } },
    { element: <Foo />, value: { id: 1 } },
    { element: <Foo />, value: { id: 2 } },
    { element: <Foo />, value: { id: 3 } },
    { element: <Foo />, value: { id: 4 } },
    { element: <Foo />, value: { id: 5 } },
    { element: <Foo />, value: { id: 6 } },
    { element: <Foo />, value: { id: 7 } },
  ];

  it('create job of multicast api', () => {
    const thread = new MulticastThread(['foo', 'bar', 'baz']);

    const jobs = thread.createJobs(actions);

    expect(jobs.length).toBe(2);
    expect(jobs).toMatchSnapshot();

    expect(jobs[0].threadId).toBe(jobs[1].threadId);

    jobs.forEach(job => {
      expect(job.entry).toBe('message/multicast');
      expect(job.body.to).toEqual(['foo', 'bar', 'baz']);
    });
  });

  it('always retrun identical threadId of jobs', () => {
    const jobs = [];

    jobs.push(...new MulticastThread(['foo', 'bar']).createJobs(actions));
    jobs.push(...new MulticastThread(['foo', 'baz']).createJobs(actions));
    jobs.push(...new MulticastThread(['bar', 'baz']).createJobs(actions));

    expect(jobs.length).toBe(6);

    const { threadId } = jobs[0];

    jobs.forEach(job => {
      expect(job.threadId).toBe(threadId);
    });
  });

  it('throw if non message action contained', () => {
    const thread = new MulticastThread(['foo', 'bar', 'baz']);

    expect(() =>
      thread.createJobs([
        ...actions.slice(0, 7),
        { element: <Bar />, value: { id: 'Boom!' } },
        actions.slice(7),
      ])
    ).toThrowErrorMatchingInlineSnapshot(
      `"<Bar /> is invalid to be delivered in multicast"`
    );
  });
});
