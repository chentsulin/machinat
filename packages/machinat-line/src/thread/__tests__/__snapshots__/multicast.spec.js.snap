// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createJobs(actions) create job of multicast api 1`] = `
Array [
  Object {
    "body": Object {
      "messages": Array [
        Object {
          "id": 0,
        },
        Object {
          "id": 1,
        },
        Object {
          "id": 2,
        },
        Object {
          "id": 3,
        },
        Object {
          "id": 4,
        },
      ],
      "to": Array [
        "foo",
        "bar",
        "baz",
      ],
    },
    "entry": "message/multicast",
    "threadId": "line:default:multicast",
  },
  Object {
    "body": Object {
      "messages": Array [
        Object {
          "id": 5,
        },
        Object {
          "id": 6,
        },
        Object {
          "id": 7,
        },
      ],
      "to": Array [
        "foo",
        "bar",
        "baz",
      ],
    },
    "entry": "message/multicast",
    "threadId": "line:default:multicast",
  },
]
`;
